File: etheroom-web/src/app/core/services/hotel-room.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environments';
import { HotelRoom } from '../models/hotel/aggregates/hotel-room.model';
import { Observable } from 'rxjs';
import { Page } from '../types/types';
import { Optional } from '../utils/optional';

@Injectable({
  providedIn: 'root'
})
export class HotelRoomService {

  private hotelRoomApiUrl: string;

  private readonly HOTEL_ROOM_PATH = 'hotel-room';

  constructor(
    private http: HttpClient
  ) {
    this.hotelRoomApiUrl = `${environment.apiUrl}/${this.HOTEL_ROOM_PATH}`;
  }

  public create(hotelRoom: HotelRoom): Observable<HotelRoom> {
    return this.http.post<HotelRoom>(this.hotelRoomApiUrl, hotelRoom);
  }

  public findAll(
    page: number,
    size: number,
    number: number,
    type: string,
    available: boolean,
    hotelId: string
  ): Observable<Page<HotelRoom>> {
    return this.http.get<Page<HotelRoom>>(
      this.hotelRoomApiUrl,
      {
        params: {
          page: page.toString(),
          size: size.toString(),
          number: Optional.ofNullable(number).map(n => n.toString()).orElse(''),
          type: Optional.ofNullable(type).orElse(''),
          available: Optional.ofNullable(available).map(a => a.toString()).orElse(''),
          hotelId: Optional.ofNullable(hotelId).orElse('')
        }
      }
    );
  }

  public findAllAvailable(
    page: number,
    size: number,
    number: number = null,
    type: string = null,
    hotelId: string = null
  ): Observable<Page<HotelRoom>> {
    return this.http.get<Page<HotelRoom>>(
      `${this.hotelRoomApiUrl}/available`,
      {
        params: {
          page: page.toString(),
          size: size.toString(),
          number: Optional.ofNullable(number).map(n => n.toString()).orElse(''),
          type: Optional.ofNullable(type).orElse(''),
          hotelId: Optional.ofNullable(hotelId).orElse('')
        }
      }
    );
  }

  public findById(id: string): Observable<HotelRoom> {
    return this.http.get<HotelRoom>(`${this.hotelRoomApiUrl}/${id}`);
  }

  public isHotelRoomBooked(id: string, checkIn: Date, checkOut: Date): Observable<boolean> {
    return this.http.get<boolean>(
      `${this.hotelRoomApiUrl}/${id}/booked`, 
      {
        params: {
          checkIn: Optional.ofNullable(checkIn).map(c => c.toISOString()).orElse(''),
          checkOut: Optional.ofNullable(checkOut).map(c => c.toISOString()).orElse('')
        }
      }
    );
  }

  public update(hotelRoom: HotelRoom): Observable<void> {
    return this.http.put<void>(this.hotelRoomApiUrl, hotelRoom);
  }

  public delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.hotelRoomApiUrl}/${id}`);
  }

}

File: etheroom-api/src/main/java/com/etheroom/Etheroom/domain/services/hotel/aggregates/HotelRoomService.java
package com.etheroom.Etheroom.domain.services.hotel.aggregates;

import com.etheroom.Etheroom.domain.models.hotel.aggregates.HotelRoom;
import com.etheroom.Etheroom.domain.repositories.hotel.aggregates.HotelRoomRepository;
import com.etheroom.Etheroom.infrastructure.utils.Functions;
import com.etheroom.Etheroom.infrastructure.vo.enums.BookingStatus;
import com.etheroom.Etheroom.infrastructure.vo.exception.exceptions.BadRequestException;
import com.etheroom.Etheroom.infrastructure.vo.exception.exceptions.NotFoundException;
import com.etheroom.Etheroom.infrastructure.vo.filter.HotelRoomFilter;
import com.etheroom.Etheroom.presentation.dtos.hotel.HotelDto;
import com.etheroom.Etheroom.presentation.dtos.hotel.aggregates.HotelRoomDto;
import com.etheroom.Etheroom.presentation.services.booking.aggregates.IBookingHelper;
import com.etheroom.Etheroom.presentation.services.hotel.IHotelService;
import com.etheroom.Etheroom.presentation.services.hotel.aggregates.IHotelRoomService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class HotelRoomService implements IHotelRoomService {

    private static final String HOTEL_ROOM_NOT_FOUND = "Hotel room not found";

    private static final String HOTEL_NOT_INFORMED = "Hotel not informed";

    private static final String ONGOING_BOOKING = "This hotel room have ongoing booking";

    private final IBookingHelper bookingHelper;

    private final IHotelService hotelService;

    private final HotelRoomRepository hotelRoomRepository;

    @Override
    public HotelRoomDto create(HotelRoomDto hotelRoomDto) {
        HotelRoom hotelRoom = hotelRoomDto.mapDtoToEntity();
        String hotelId = Optional.ofNullable(hotelRoomDto.getHotelId())
                .map(Objects::toString)
                .orElseThrow(() -> new BadRequestException(HOTEL_NOT_INFORMED));
        HotelDto hotelDto = this.hotelService.findById(hotelId);
        hotelRoom.setHotel(hotelDto.mapDtoToEntity());
        return this.hotelRoomRepository.save(hotelRoom).mapEntityToDto();
    }

    @Override
    public Page<HotelRoomDto> findAll(Pageable pageable, HotelRoomFilter filter) {
        return this.hotelRoomRepository.findAll(
                        pageable,
                        filter.getNumber(),
                        filter.getType(),
                        filter.getHotelId(),
                        filter.getAvailable()
                ).map(HotelRoom::mapEntityToDto);
    }

    @Override
    public Page<HotelRoomDto> findAllAvailable(Pageable pageable, HotelRoomFilter filter) {
        return this.hotelRoomRepository.findAll(
                        pageable,
                        filter.getNumber(),
                        filter.getType(),
                        filter.getHotelId(),
                        Boolean.TRUE
                ).map(HotelRoom::mapEntityToDto);
    }

    @Override
    public HotelRoomDto findById(String id) {
        return this.hotelRoomRepository.findById(UUID.fromString(id))
                .map(HotelRoom::mapEntityToDto)
                .orElseThrow(() -> new NotFoundException(HOTEL_ROOM_NOT_FOUND));
    }

    @Override
    public Boolean isHotelRoomBooked(String roomId, LocalDateTime checkIn, LocalDateTime checkOut) {
        return this.bookingHelper.isHotelRoomBooked(roomId, checkIn, checkOut);
    }

    @Override
    public void update(HotelRoomDto hotelRoomDto) {
        UUID uuid = hotelRoomDto.getId();
        Functions.acceptFalseThrows(
                hotelRoomRepository.existsById(uuid),
                () -> new NotFoundException(HOTEL_ROOM_NOT_FOUND)
        );
        HotelRoom hotelRoom = hotelRoomDto.mapDtoToEntity();
        this.hotelRoomRepository.save(hotelRoom);
    }

    @Override
    public void delete(String id) {
        UUID uuid = UUID.fromString(id);
        Functions.acceptTrueThrows(
                hotelRoomRepository.existsById(uuid),
                () -> new NotFoundException(HOTEL_ROOM_NOT_FOUND)
        );
        Functions.acceptTrueThrows(
                this.bookingHelper.existsByRoomIdAndStatus(id, BookingStatus.ACTIVE),
                () -> new BadRequestException(ONGOING_BOOKING)
        );
        this.hotelRoomRepository.deleteById(uuid);
    }
}
